### DISCUSSION TITLE
Button click from code > popup open in new window instead of new tab

### DISCUSSION THREAD

**Discussion by kingchenc** [cefsharp/CefSharp#4099] [General]
Hello, 

im bit struggeling about correct handling new windows from a popup with a button click.

I know we can use LifeSpanHandler, thats what i have done at the moment.

this opens a new tab with the url that i set.

```

               chromeBrowser = new ChromiumWebBrowser("google.com");
               LifespanHandler life = new LifespanHandler();
               chromeBrowser.LifeSpanHandler = life;
               life.popup_request += life_popup_request;
               this.tabControl1.TabPages.Add(chromeBrowser);
               chromeBrowser.Dock = DockStyle.Fill;
```

other test with OnBeforePopup...

```
         public bool OnBeforePopup(IWebBrowser chromiumWebBrowser, IBrowser browser, IFrame frame, string targetUrl, string targetFrameName, WindowOpenDisposition targetDisposition, bool userGesture, IPopupFeatures popupFeatures, IWindowInfo windowInfo, IBrowserSettings browserSettings, ref bool noJavascriptAccess, out IWebBrowser newBrowser)
            {
                browser.MainFrame.LoadUrl(targetUrl);
                newBrowser = null;
                return true;

            }
```
this code is only working, if i provide a URL.

but my code set the attribute, and then click it.
yes this working well, but the popup after the click opens in new window, but i need to get a new tab.

```

            chromeBrowser.ExecuteScriptAsync("document.querySelector('.btn3').setAttribute('target', '_blank');");

            chromeBrowser.ExecuteScriptAsync("document.querySelector('.btn3').click();");
```

Anyone know how is that possible to make the ExecuteScriptAsync............click(); opens in a new tab?

**Comment by amaitland** [comment]
In future please use discussions to ask questions. https://github.com/cefsharp/CefSharp/discussions

If you are using WinForms then see 
https://github.com/cefsharp/CefSharp/pull/3529#issuecomment-828934239

There is a working example as part of the project source see https://github.com/cefsharp/CefSharp/blob/cefsharp/101/CefSharp.WinForms.Example/BrowserTabUserControl.cs#L90

**Comment by kingchenc** [comment]
Hey,

im sorry, was the wrong subforum :( @amaitland 

thanks for the links, I think i get worked, just modified my code a bit :) 

```
               LifespanHandler life = new LifespanHandler();
               chromeBrowser.LifeSpanHandler = life;
               life.popup_request += life_popup_request;
               chromeBrowser.ExecuteScriptAsync("document.querySelector('.btn3').click();");
```

This will load the popup in a new tab! Nice :)

---

Calling this after the new tab from the popup:

                `chromeBrowser.ExecuteScriptAsync("document.getElementsById('.testbutton').click();");`


`INFO:CONSOLE(1)] "Uncaught TypeError: document.getElementsById is not a function", source: about:blank (1)`

what do i wrong here at this point after the new popup is created as a new tab and what im missing? (yes, i have the correct Id for the testbutton)

**Reply by amaitland** [reply]
> INFO:CONSOLE(1)] "Uncaught TypeError: document.getElementsById is not a function", source: about:blank (1)

Fix your JavaScript. 

https://developer.mozilla.org/en-US/docs/Web/API/Document/getElementById

**Comment by kingchenc** [comment]
Yes thats true, but i dont think thats the solution for that ..

`Uncaught TypeError: Cannot read properties of null (reading 'click')", source: about:blank (1)`

`chromeBrowser.ExecuteScriptAsync("document.querySelector('.btn3').click();");`
(same code works on the mainpage after loading the url, but not on a new created popup tab)

Do i need after the popup is created in a new tab to set focus, handler, Register or so?
seems for me that the script dont have access to the new tab.

tryed this in the OnAfterCreated class:

                browser.GetHost();
                browserControl.GetBrowserHost();
                browserControl.ResourceRequestHandlerFactory = new ResourceRequestHandlerFactory();

any working example? :/

**Reply by amaitland** [reply]
> Yes thats true, but i dont think thats the solution for that ..

Solution for what exactly?

> (same code works on the mainpage after loading the url, but not on a new created popup tab)

What are you trying to achieve exactly? Please be clear and concise, takes a few minutes and review your answer before posting please.

> any working example? :/

It's not clear what you are asking, so I cannot provide an example or guidance.

Start from the beginning, clearly outline what it is you are trying to achieve.

**Reply by amaitland** [reply]
Please read over the following as they will hopefully give you some guidance on how to clearly express what it is you are trying to achieve:

- https://codeblog.jonskeet.uk/2012/11/24/stack-overflow-question-checklist/
- https://github.com/cefsharp/CefSharp/wiki/Questions-&-Support#when-asking-a-question-please-include-the-following

**Comment by kingchenc** [comment]
Okay Sir,
i will try again, im sorry, English is not my natural language.

My code opens a website, after that i login to my usercenter.
Then it clicks on a button, this opens the popup in a new tab.
After that, i need to click again a button, but i struggle with it.

Code:

```
            CefSettings settings = new CefSettings();
            Cef.Initialize(new CefSettings { CachePath = Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData) + @"\CEF", PersistSessionCookies = true });

            string cache_dir = Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData) + @"\CEF";
            settings.CachePath = cache_dir;
            settings.CefCommandLineArgs.Add("persist_session_cookies", "1");
            settings.LogSeverity = LogSeverity.Disable;

            string link = "URL";

            chromeBrowser = new ChromiumWebBrowser(link);

            TabPage tp = new TabPage();
            tp.Text = @"SomeText";
            tp.Controls.Add(chromeBrowser);
            tabControl1.TabPages.Add(tp);
            
            chromeBrowser.Dock = DockStyle.Fill;

            chromeBrowser.Load(link);


            ..... (login to my user center)

            LifespanHandler life = new LifespanHandler();
            chromeBrowser.LifeSpanHandler = life;
            life.popup_request += life_popup_request;

            chromeBrowser.Load(link1);
           
           // click button to open the popup in tab
           chromeBrowser.ExecuteScriptAsync("document.querySelector('.btn3').click();");

          //... some wait functions here
          //want to click here on the new tab a button 
          chromeBrowser.ExecuteScriptAsync("document.querySelector('.btn3').click();");

```




```
        public class LifespanHandler : ILifeSpanHandler
        {
            //event that receive url popup
            public event Action<string> popup_request;

            bool ILifeSpanHandler.OnBeforePopup(IWebBrowser browserControl, IBrowser browser, IFrame frame, string targetUrl, string targetFrameName, WindowOpenDisposition targetDisposition, bool userGesture, IPopupFeatures popupFeatures, IWindowInfo windowInfo, IBrowserSettings browserSettings, ref bool noJavascriptAccess, out IWebBrowser newBrowser)
            {
                //get url popup
                if (this.popup_request != null)
                    this.popup_request(targetUrl);

                //stop open popup
                newBrowser = null;
                return true;
            }

            bool ILifeSpanHandler.DoClose(IWebBrowser browserControl, IBrowser browser)
            { return false; }


            void ILifeSpanHandler.OnBeforeClose(IWebBrowser browserControl, IBrowser browser) { }


            void ILifeSpanHandler.OnAfterCreated(IWebBrowser browserControl, IBrowser browser) 
            {
                browser.GetHost();
                browserControl.GetBrowserHost();
                browserControl.ResourceRequestHandlerFactory = new ResourceRequestHandlerFactory();
            }
        }

        private void life_popup_request(string popup_request)
        {
            this.Invoke((MethodInvoker)delegate () {
                string title = "TabPage " + (tabControl1.TabCount + 1).ToString();

                TabPage myTabPage = new TabPage(title);

                tabControl1.TabPages.Add(myTabPage);

                tabControl1.SelectTab(tabControl1.TabCount - 1);

                ChromiumWebBrowser chromeBrowser = new ChromiumWebBrowser(popup_request);

                chromeBrowser.LifeSpanHandler = life;

                chromeBrowser.Parent = myTabPage;

                chromeBrowser.Dock = DockStyle.Fill;
            });
        }
```

This works for the Mainpage, but not on the new created Tab.
`chromeBrowser.ExecuteScriptAsync("document.querySelector('.btn3').click();");`

`Uncaught TypeError: Cannot read properties of null (reading 'click')", source: about:blank (1)`

And yes, the querySelector is correct for the website in the new tab.
I hope, my question is now clear.

**Reply by amaitland** [reply]
> After that, i need to click again a button, but i struggle with it.

Where does this button exist exactly? In the popup?

Are you executing your code against the correct `ChromiumWebBrowser` instance? I cannot tell from your code, likely this is your problem.

**Reply by amaitland** [reply]
Another possibility is you are executing your script before the DOM has loaded.
